Deprecation guide (#1018)
conda-incubator/conda-store: docusaurus-docs/community/policies/backwards-compatibility.md

#### Removing versions of API endpoints

Removal of an API endpoint is sometimes necessary. Deprecation notices and removals will
always be documented in the [conda-store release notes](https://github.com/conda-incubator/conda-store/blob/main/CHANGELOG.md). Further, documentation should be updated to reflect these changes. This should include:
- version number of the release where the deprecation was introduced
- provide suggestions for alternatives (if possible)
- provide justification for the removal (such as a link to the issue
  or CVE that necessitated the removal).

In order to make these kind of breaking changes responsibly, follow the steps
outlined below.

##### 1. Deprecate the endpoint

In this stage, conda-store should not be introducing a breaking change. The
deprecation step communicates to users that the endpoint is marked for removal

> and will removed in a future release.

Reviewer: ```suggestion
and will be removed in a future release.
```

#### Removing versions of API endpoints

Removal of an API endpoint is sometimes necessary. Deprecation notices and removals will
always be available in [release notes](https://github.com/conda-incubator/conda-store/blob/main/CHANGELOG.md). Further, documentation should be updated to reflect these changes. This should include:
- version number of the release where this was deprecated
- provide suggestions for alternatives (if possible)
- provide justification for the removal (such as a link to the issue
  or CVE that necessitated the removal).

In order to make these kind of breaking changes responsibly, follow the steps
outlined below.

##### 1. Deprecate the endpoint

In this stage, conda-store should not be introducing a breaking change. The
deprecation step communicates to users that the endpoint is marked for removal

> and will removed in a future release.

Reviewer: We must establish a timeline for deprecation/sunsetting to set expectations accordingly.

> and will removed in a future release.

Since we use CalVer, we should leave _at least_ two/three months before removing any endpoints; otherwise, "a future release" can be anything between a couple of weeks to an indeterminate amount.

:::info[For vulnerabilities]

In the case of a serious security vulnerability, conda-store may skip the deprecation
step and remove the endpoint immediately.

:::

Deprecations in the REST API follow the
form outlined by the [deprecation header RFC](https://datatracker.ietf.org/doc/html/draft-ietf-httpapi-deprecation-header-02). To deprecate an endpoint add the following

> response headers to the endpoint
> 
> ```
> {
>    "Deprecation": "True",
>    "Sunset": <removal date, eg. "Mon, 16 Feb 2025 23:59:59 UTC" >
> }
> ```
> 
> The "removal date" indicates the date after which conda store will no longer serve

Reviewer: ```suggestion
response headers to the endpoint:

```json
{
   "Deprecation": "True",
   "Sunset": <removal date, eg. "Mon, 16 Feb 2025 23:59:59 UTC" >
}
```

The `removal date` indicates the date after which conda-store will no longer serve
```

the endpoint. It will be specified as a [HTTP-Date](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Date). The removal date can be set to (at earliest) the date of the target

> next release.

Reviewer: See comment above:

> The removal date can be set to (at earliest) the date of the target
next release.

Since we use CalVer and _do not_ have set release dates, this puts us in a position where we need to agree on a release and meet it no matter what.
Since we do not have a set release schedule, this is rather ambiguous, so we need a clearer guideline here.

Author: I can be more specific here. 

I think sunset date should be set to about 2 months from the release that the deprecation is going out in. 
* Any time before this date, users should expect this endpoint to work
* Any time after this date, the endpoint may still be available (due to no release having gone out), but users should expect that this endpoint will disappear

For example, say you have an endpoint `/hello-planet` that is getting deprecated. The whole deprecation cycle will be something like:
* 2024/12/17 - add deprecation headers to `/hello-planet`,  sunset date is `2025/03/17`
* 2024/12/19 - new release goes out (sunset date is more than 2 months away, so everything is good)
* 2025/01/30 - new release goes out, `/hello-planet` is still deprecated
* 2025/03/16 - new emergency release goes out to fix a discovered security vulnerability. Still before the `/hello-planet` sunset date, so it remains in the deprecated state
* 2025/03/17 - `/hello-planet` sunset date. No release so this endpoint is still functional, however will be yanked at next release
* 2025/03/29 - new release goes out `/hello-planet` endpoint functionality is removed, returns a `410` response with removal date (eg. 2025/04/29)
* 2025/04/29 - new release goes out. Happens to coincide with the removal date. Can remove the endpoint as part of this release

Author: I don't think the removal of functionality step offers much more help to users than the deprecation step. If the team is open to changing this more, I would advocate for jumping straight from deprecating the endpoint to removing the endpoint after the sunset date.

##### 2. Remove endpoint functionality

At lease one release from the deprecation time, the functionality for the endpoint

> can be removed. To indicate that this endpoint is no longer functional, the endpoint

Reviewer: Same comment about timeline

must:
* return a status code of `410 Gone`
* return a json object stating when and why the endpoint was removed and what
version of the endpoint is available currently (if any).

> 
> ```python

Reviewer: ```suggestion

For example:

```python
```

{
  # the pull request that removed the endpoint
  "reference_pull_request": "https://github.com/conda-incubator/conda-store/pull/0000",
  # the date the endpoint was removed (as HTTP-Date). This should be after the Sunset date specified
  # in the deprecation step
  "removal_date": "Mon, 16 Feb 2025 23:59:59 UTC"
  # the reason for the removal, ideally with a link to a CVE if one is available
  "removal_reason": "Removed to address CVE-2021-32677 (https://nvd.nist.gov/vuln/detail/CVE-2021-32677)",
  # the endpoint that developers should use as a replacement (if applicable)
  "new_endpoint": "api/v3/this/should/be/used/instead",
}
```

##### 3. Remove the endpoint

At least one release from the completion of step (2), the API endpoint may be fully
removed. At this stage, users should expect to recieve a `404 Not Found` error for the

> endpoint.

Reviewer: Same comment about timelines. 
I know this is in the original doc, but with our release variation we need to be stricter about this.

#### Removing versions of API endpoints

Removal of an API endpoint is sometimes necessary. Deprecation notices and removals will

> always be available in [release notes](https://github.com/conda-incubator/conda-store/blob/main/CHANGELOG.md). Further, documentation should be updated to reflect these changes. This should include:

Reviewer: ```suggestion
always be documented in the [conda-store release notes](https://github.com/conda-incubator/conda-store/blob/main/CHANGELOG.md). Further, documentation should be updated to reflect these changes. This should include:
```

> - version number of the release where this was deprecated

Reviewer: ```suggestion
- version number of the release where the deprecation was introduced
```

- provide suggestions for alternatives (if possible)
- provide justification for the removal (such as a link to the issue
  or CVE that necessitated the removal).

In order to make these kind of breaking changes responsibly, follow the steps
outlined below.

##### 1. Deprecate the endpoint

In this stage, conda-store should not be introducing a breaking change. The
deprecation step communicates to users that the endpoint is marked for removal
and will removed in a future release.

:::info[For vulnerabilities]

In the case of a serious security vulnerability, conda-store may skip the deprecation
step and remove the endpoint immediately.

:::

Deprecations in the REST API follow the
form outlined by the [deprecation header RFC](https://datatracker.ietf.org/doc/html/draft-ietf-httpapi-deprecation-header-02). To deprecate an endpoint add the following
response headers to the endpoint

```json
{
   "Deprecation": "True",
   "Sunset": <removal date, eg. "Mon, 16 Feb 2025 23:59:59 UTC" >
}
```

The "removal date" should be specified as a [HTTP-Date](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Date).

###### Choosing a removal date

> The `Sunset` date indicates the date after which conda store with remove the functionality

Reviewer: ```suggestion
The `Sunset` date indicates the date after which conda-store will remove the functionality
```

from the release that the deprecation notice first appears in.

##### 2. Remove endpoint functionality

Once we have reached the `Sunset` date (from the step above), conda-store may
remove the functionality of the endpoint. To indicate that this endpoint is no
longer functional, the endpoint must:
* return a status code of `410 Gone`
* return a json object stating when and why the endpoint was removed and what
version of the endpoint is available currently (if any).

For example:

```python
{
  # the pull request that removed the endpoint
  "reference_pull_request": "https://github.com/conda-incubator/conda-store/pull/0000",
  # the date the endpoint was removed (as HTTP-Date). This should be after the Sunset date specified
  # in the deprecation step
  "removal_date": "Mon, 16 Feb 2025 23:59:59 UTC"
  # the reason for the removal, ideally with a link to a CVE if one is available
  "removal_reason": "Removed to address CVE-2021-32677 (https://nvd.nist.gov/vuln/detail/CVE-2021-32677)",
  # the endpoint that developers should use as a replacement (if applicable)
  "new_endpoint": "api/v3/this/should/be/used/instead",
}
```

> ###### Choosing a removal date

Reviewer: This is a repeat of the section starting in line 178, which needs reviewing.

The `removal_date` date indicates the date after which conda store with remove the endpoint
from the codebase. After this point, the endpoint will return a `404` response code.
* Any time before this date, users should expect this endpoint to return a `410` response code

> * Any time after this date, the endpoint may still return a `410` response (due to no release having gone out), but users should expect that this endpoint will disappear

Reviewer: I am having trouble following these points. Per the above (line 190)

> To indicate that this endpoint is no
longer functional, the endpoint must:
* return a status code of `410 Gone`
* return a json object stating when and why the endpoint was removed and what
version of the endpoint is available currently (if any).

Once we reach the `sunset` (or removal date), the endpoint should return a `410`, but the bullets here seem to indicate that the endpoint should return a `410` _before_ the `sunset` date and a `404` after, which contradicts the steps above.

These two sections need aligning. To keep things simple we should just stick to a `410` I think,

Since the conda-store project users CalVer, the `removal_date` date should be set to at least 2 months
from the release that the deprecation notice first appears in.

##### 3. Remove the endpoint

Once we have reached the `removal_date` from step (2), the API endpoint may be fully
removed. At this stage, users should expect to recieve a `404 Not Found` error for the

> endpoint.

Reviewer: Also a repeated section, needs review

Public variables should not have their type changed.

Public constants should not have their type or their value changed.

> ##### Config

Reviewer: ```suggestion
##### Configuration elements
```

> Config elements that are deprecated will be marked with a `deprecation` note

Reviewer: ```suggestion
Configuration elements that are deprecated will be marked with a `deprecation` note
```

Public variables should not have their type changed.

Public constants should not have their type or their value changed.

##### Configuration elements

> Configuration elements that are deprecated will be marked with a `deprecation` note
> in the docs and `--help` output. For example:
> ```
> $ conda-store-server --help-all
> 

Reviewer: ```suggestion
Deprecated configuration elements will be marked with a `deprecation` note
in the docs and the CLI `--help` output. For example:
```
$ conda-store-server --help-all

```
